Assets/Scripts/NodeGameObjectFactory.cs

```diff
@@ -1,16 +1,62 @@
-﻿using UnityEngine;
+﻿using Assets.Scripts.Common;
+using Assets.Scripts.Model;
+using UnityEngine;
 
 namespace Assets.Scripts
 {
   public class NodeGameObjectFactory : MonoBehaviour
   {
+    TextureCache _materialCache;
+
     public GameObject nodePrefab;
     public GameObject ghostNodePrefab;
+    public GameObject nodeCanvasPrefab;
+
+    const PrimitiveType c_defaultNodeType = PrimitiveType.Sphere;
+
+    public void Start()
+    {
+      _materialCache = FindObjectOfType<TextureCache>();
+    }
+
+
+    public GameObject CreateNodeGameObject(Node node)
+    {
+      if (_materialCache == null)
+        _materialCache = FindObjectOfType<TextureCache>();
+
+      if (node.nodeClass == null)
+        return CreateDefaultNode(node.Position, Quaternion.Euler(node.Rotation));
+
+      var material = _materialCache.GetByClassId(node.nodeClass.id);
+
+      var gameObject = node.nodeClass.shape == null
+        ? GameObject.CreatePrimitive(c_defaultNodeType)
+        : GameObject.CreatePrimitive(node.nodeClass.shape.Value);
+
+      gameObject.transform.localScale = Vector3.one * (node.nodeClass.scale ?? 1);
+      gameObject.transform.position = node.Position;
+      gameObject.transform.rotation = Quaternion.Euler(node.Rotation);
+      gameObject.tag = Constants.PhysicalNodeTag;
+
+      var renderer = gameObject.GetComponent<Renderer>();
+      renderer.material = material;
+
+      Instantiate(nodeCanvasPrefab, gameObject.transform);
+
+      return gameObject;
+    }
 
-    public GameObject CreateNodeGameObject(Vector3 position, Quaternion rotation) =>
-      Instantiate(nodePrefab, position, rotation);
 
     public GameObject CreateGhostNodeGameObject(Vector3 position, Quaternion rotation) =>
       Instantiate(ghostNodePrefab, position, rotation);
+
+    public GameObject CreateDefaultNode(Vector3 position, Quaternion rotation) =>
+      Instantiate(nodePrefab, position, rotation);
+
+    void Log(string message)
```

Comment:
why?

Reply:
This is temporary, added because debugging was slow. I will still work on this, so I did not refine the initial version that much.

Ideally I want to make use of an ILogger and use that instead. Still to consider how the logs should be displayed to the user (custom window? toasts?)

Reply:
we already have toasts, so i would stick with them

Reply:
Toasts are great for displaying quick messages but are not fit for longer logs. In case of json validation, ideally we would want to display all errors in it, and that might be multiple lines. That's why I'm considering a transparent window for messages such as those. 
Besides, Toasts only display for a few seconds. logs should be visible to the user on demand.